File: apps/backend-gateway/src/infrastructure/repositories/gate_repository.py
"""
StadiumOS AI — Gate Repository Adapter.

SQLAlchemy implementation of the GateRepository port.
"""

from uuid import UUID
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.application.ports.gate_repository import GateRepository
from src.domain.entities.gate import Gate
from src.domain.value_objects import Coordinates
from src.infrastructure.database.models.gate_model import GateModel


class SqlGateRepository(GateRepository):
    """
    SQLAlchemy implementation of the GateRepository interface.
    """

    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    def _to_domain(self, model: GateModel) -> Gate:
        """Map ORM model to domain entity."""
        return Gate(
            id=model.id,
            sector_id=model.sector_id,
            gate_code=model.gate_code,
            location=Coordinates(latitude=model.latitude, longitude=model.longitude),
            status=model.status,
            is_bidirectional=model.is_bidirectional,
            created_at=model.created_at,
            updated_at=model.updated_at,
        )

    async def get_by_id(self, gate_id: UUID) -> Gate | None:
        stmt = select(GateModel).where(GateModel.id == gate_id)
        result = await self.session.execute(stmt)
        model = result.scalar_one_or_none()
        return self._to_domain(model) if model else None

    async def get_by_code(self, gate_code: str) -> Gate | None:
        stmt = select(GateModel).where(GateModel.gate_code == gate_code.upper().strip())
        result = await self.session.execute(stmt)
        model = result.scalar_one_or_none()
        return self._to_domain(model) if model else None

    async def list_by_sector(self, sector_id: UUID) -> list[Gate]:
        stmt = select(GateModel).where(GateModel.sector_id == sector_id).order_by(GateModel.gate_code)
        result = await self.session.execute(stmt)
        return [self._to_domain(m) for m in result.scalars().all()]

    async def save(self, gate: Gate) -> Gate:
        stmt = select(GateModel).where(GateModel.id == gate.id)
        result = await self.session.execute(stmt)
        model = result.scalar_one_or_none()

        if model:
            model.sector_id = gate.sector_id
            model.gate_code = gate.gate_code
            model.status = gate.status
            model.latitude = gate.location.latitude
            model.longitude = gate.location.longitude
            model.is_bidirectional = gate.is_bidirectional
            model.updated_at = gate.updated_at
        else:
            model = GateModel(
                id=gate.id,
                sector_id=gate.sector_id,
                gate_code=gate.gate_code,
                status=gate.status,
                latitude=gate.location.latitude,
                longitude=gate.location.longitude,
                is_bidirectional=gate.is_bidirectional,
                created_at=gate.created_at,
                updated_at=gate.updated_at,
            )
            self.session.add(model)

        await self.session.flush()
        return self._to_domain(model)

    async def delete(self, gate_id: UUID) -> None:
        stmt = select(GateModel).where(GateModel.id == gate_id)
        result = await self.session.execute(stmt)
        model = result.scalar_one_or_none()
        if model:
            await self.session.delete(model)
            await self.session.flush()

File: apps/backend-gateway/src/infrastructure/repositories/gateRepository.ts
/**
 * StadiumOS AI — Gate Repository Adapter.
 * 
 * Implements IGateRepository using Prisma Client.
 */

import { IGateRepository } from "../../application/ports/gateRepository";
import { Gate } from "../../domain/entities/gate";
import { GateStatus } from "../../domain/enums";
import { Coordinates } from "../../domain/valueObjects";
import { prisma } from "../database/prisma";
import { Gate as PrismaGateModel } from "@prisma/client";

export class PrismaGateRepository implements IGateRepository {
  private _toDomain(model: PrismaGateModel): Gate {
    return new Gate(
      model.id,
      model.sector_id,
      model.gate_code,
      new Coordinates(model.latitude, model.longitude),
      model.status as GateStatus,
      model.is_bidirectional,
      model.created_at,
      model.updated_at
    );
  }

  public async getById(id: string): Promise<Gate | null> {
    const model = await prisma.gate.findUnique({
      where: { id }
    });
    return model ? this._toDomain(model) : null;
  }

  public async getByCode(code: string): Promise<Gate | null> {
    const model = await prisma.gate.findUnique({
      where: { gate_code: code.trim().toUpperCase() }
    });
    return model ? this._toDomain(model) : null;
  }

  public async listBySector(sectorId: string): Promise<Gate[]> {
    const models = await prisma.gate.findMany({
      where: { sector_id: sectorId },
      orderBy: { gate_code: "asc" }
    });
    return models.map((m) => this._toDomain(m));
  }

  public async save(gate: Gate): Promise<Gate> {
    const data = {
      sector_id: gate.sector_id,
      gate_code: gate.gate_code,
      status: gate.status,
      latitude: gate.location.latitude,
      longitude: gate.location.longitude,
      is_bidirectional: gate.is_bidirectional,
      updated_at: gate.updatedAt
    };

    const model = await prisma.gate.upsert({
      where: { id: gate.id },
      update: data,
      create: {
        id: gate.id,
        ...data,
        created_at: gate.created_at
      }
    });

    return this._toDomain(model);
  }

  public async delete(id: string): Promise<void> {
    await prisma.gate.delete({
      where: { id }
    });
  }
}
export default PrismaGateRepository;
